CMakeLists.txt

```diff
@@ -1,23 +1,65 @@
 cmake_minimum_required(VERSION 3.12) #Python3 package locator only added in 3.12
 
+project(cmakedoc DESCRIPTION "CMake automatic documentation generator" LANGUAGES NONE) #Not using any compiled languages so don't list, removes check for compilers
+
 find_package(Python3 COMPONENTS Interpreter)
 
+#[[[
+# Custom function to locate a Python module. Requires the Python interpreter to be located first.
+#
+# Sets PY_${module_upper} to the module file's location if found.
+#
+# :param module: Python module name to locate
+# :param REQUIRED: Optional paramater, specifies that this module is required for continued execution.
+# Sets ${module}_FIND_REQUIRED to true.
+#
+#]]
+function(find_python_module module)
+	string(TOUPPER ${module} module_upper)
+	if(NOT PY_${module_upper})
+		if(ARGC GREATER 1 AND ARGV1 STREQUAL "REQUIRED")
+			set(${module}_FIND_REQUIRED TRUE)
+		endif()
+		# A module's location is usually a directory, but for binary modules
+		# it's a .so file.
+		execute_process(COMMAND "${Python3_EXECUTABLE}" "-c" 
+			"import re, ${module}; print(re.compile('/__init__.py.*').sub('',${module}.__file__))"
+			RESULT_VARIABLE _${module}_status 
+			OUTPUT_VARIABLE _${module}_location
+		        ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+		if(NOT _${module}_status)
+			set(PY_${module_upper} ${_${module}_location} CACHE STRING 
+				"Location of Python module ${module}")
+		endif(NOT _${module}_status)
+	endif(NOT PY_${module_upper})
+	find_package_handle_standard_args(PY_${module} DEFAULT_MSG PY_${module_upper})
+endfunction(find_python_module)
+
+
 if (Python3_FOUND)
-    set(SETUP_PY_IN "${CMAKE_CURRENT_SOURCE_DIR}/setup.py") #setup.py has logic to automatically switch between CMake variables and normal operation
-    set(SETUP_PY    "${CMAKE_CURRENT_BINARY_DIR}/setup.py")
-    set(DEPS        "${CMAKE_CURRENT_SOURCE_DIR}/cmakedoc/__init__.py")
-    set(OUTPUT      "${CMAKE_CURRENT_BINARY_DIR}/build/timestamp")
+    #We have Python, now let's check to see if we have setuptools
+    find_python_module(setuptools REQUIRED)
+
+    if (PY_SETUPTOOLS)
```

Comment:
In theory `find_package_handle_standard_args` should take care of the `REQUIRED` for you. In practice I remember it being touchy.

Reply:
Yes, there's a small hack here:

https://github.com/AutonomicPerfectionist/CMakeDoc/blob/f216896aa7a0e04b07468c28521bb1922aebf421/cmake/find_python_module.cmake#L20

This should trick `find_package_handle_standard_args` into thinking it was called via the normal `find_package(<module> REQUIRED)` method and handle it accordingly. I haven't discovered any bugs so far, and converting it to a true Find Module would remove that line and let find_package handle it. However, from what I understand a Find Module is tailored specifically for a particular package, so this module would have to be hard-coded for setuptools. Hence why a more general function approach here, it should allow greater flexibility in the future.

Reply:
I agree that your general solution is better than writing specific `FindXXX` modules. The comment was meant more so that you could remove the "if" statement if you wanted.